sdk/iOS/test/MSSyncTableTests.m

```diff
@@ -1082,9 +1083,348 @@ -(void) testPushNetworkTimeoutWithErroredOps
     XCTAssertEqual(filteredClient.syncContext.pendingOperationsCount, 3);
 }
 
+// Tests cancellation of the push operation at various stages of a push workflow.
+// - Verifies that a push following a cancelled push operation works as expected
+// - Verifies that cancelling a push operation finishes the push NSOperation regardless of the cancellation point
+// - Verifies that the expected number of items are pushed to the server
+-(void) testPushCancellability
```

Comment:
OK, I am planning of breaking down the push test into the following (and similar for pull). This way there are 2 tests:
- testpush_200: a successful push that gets cancelled at various possible points during its execution
- testpush_500: an unsuccessful push that gets cancelled at various possible points during its execution

Shared code will continue to have assertions that should hold true irrespective of the cancellation point and error codes.

-(void) testPushCancellability_200
{
    XCTAssertEqual([self performPushWithCancellationPoint:0 serverResponseCode:@200], 0, "Push cancellation failed");
    XCTAssertEqual([self performPushWithCancellationPoint:1 serverResponseCode:@200], 0, "Push cancellation failed");
    XCTAssertEqual([self performPushWithCancellationPoint:2 serverResponseCode:@200], 1, "Push cancellation failed");
    XCTAssertEqual([self performPushWithCancellationPoint:3 serverResponseCode:@200], 0, "Push cancellation failed");
}

-(void) testPushCancellability_500
{
    XCTAssertEqual([self performPushWithCancellationPoint:0 serverResponseCode:@500], 0, "Push cancellation failed");
    XCTAssertEqual([self performPushWithCancellationPoint:1 serverResponseCode:@500], 0, "Push cancellation failed");
    XCTAssertEqual([self performPushWithCancellationPoint:2 serverResponseCode:@500], 0, "Push cancellation failed");
    XCTAssertEqual([self performPushWithCancellationPoint:3 serverResponseCode:@500], 0, "Push cancellation failed");
}